lib/src/serializable.dart

```diff
@@ -0,0 +1,224 @@
+      <String, _ModuleRegistration>{};
+
+  Bridge get bridge => _bridge;
+  set bridge(Bridge bridge) {
+    _bridge = bridge;
+    _bridge.apiCallReceived.listen(_handleApiCall);
+  }
+
+  Map<String, SerializableModule> get registeredModules =>
+      new Map.fromIterable(_moduleRegistrations.keys,
+          value: (key) => _moduleRegistrations[key].module);
+
+  void reset() {
+    _moduleRegistrations.clear();
+    _bridge = null;
+  }
+
+  void deregisterModule(SerializableModule module) {
+    if (module.serializableKey == null) return;
+    if (_moduleRegistrations.containsKey(module.serializableKey)) {
+      _moduleRegistrations[module.serializableKey].dispose();
+      _moduleRegistrations.remove(module.serializableKey);
+    }
+  }
+
+  void registerModule(SerializableModule module) {
+    if (module.serializableKey != null) {
+      final registration = new _ModuleRegistration(module);
+      _registerForLifecycleEvents(registration);
+      _moduleRegistrations[module.serializableKey] = registration;
+    } else {
+      _logger.warning('Unable to serialize module without serializableKey');
+    }
+  }
+
+  void _handleApiCall(Map apiCall) {
+    String module = apiCall['module'];
+    String method = apiCall['method'];
+    Object data = apiCall['data'];
+
+    SerializableModule targetModule = _moduleRegistrations[module].module;
+
+    if (targetModule != null && data is List) {
+      _deserializeAndCall(targetModule, method, data);
+    }
+  }
+
+  void _deserializeAndCall(
+      SerializableModule module, String method, List data) {
+    InstanceMirror apiMirror = reflect(module.api);
+
+    if (apiMirror == null) {
+      _logger.warning(
+          'Unable to create mirror on api for ${module.serializableKey}');
+      return;
+    }
+
+    ClassMirror classMirror = apiMirror.type;
+    MethodMirror apiMethodMirror = classMirror.declarations[new Symbol(method)];
+
+    if (apiMethodMirror == null) {
+      _logger.warning(
+          'Method $method does not exist on ${module.serializableKey}\' module\'s API');
+      return;
+    }
+
+    // Check here that the position args in data match the expected params of the method being called
+    if (apiMethodMirror.parameters.length == data.length) {
+      for (var i = 0; i < apiMethodMirror.parameters.length; i++) {
+        var param = apiMethodMirror.parameters[i];
+
+        if (data[i] is Map && param.type.reflectedType != Map) {
+          ClassMirror paramClassMirror = reflectClass(param.type.reflectedType);
+
+          // Paramter type must implement fromJson name constructor that takes a Map
+          try {
+            var instance = paramClassMirror
+                .newInstance(new Symbol('fromJson'), [data[i]]).reflectee;
+            data[i] = instance;
+          } on NoSuchMethodError {
+            _logger.warning(
+                '${paramClassMirror.simpleName.toString()} does not implement fromJson named constructor');
+            return;
```

Comment:
I think it is useful because we get a meaningful error message instead of the error being thrown in the console.